[GEP-30] Rework API Server Proxy to Drop Proxy Protocol (#10793)
gardener/gardener: docs/proposals/30-apiserver-proxy.md

## Table of Contents

- [Summary](#summary)
- [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
- [Proposal](#proposal)
- [Current Architecture](#current-architecture)
- [Proposed Changes](#proposed-changes)
- [Custom Header Specification](#custom-header-specification)
- [HTTP CONNECT Implementation](#http-connect-implementation)
    - [Technical Implementation Details](#technical-implementation-details)
    - [Istio IngressGateway Configuration](#istio-ingressgateway-configuration)
    - [EnvoyFilter for Custom Header Processing](#envoyfilter-for-custom-header-processing)
    - [Implementation Steps](#implementation-steps)
- [Feature Gate Implementation](#feature-gate-implementation)
    - [Feature Gate 1: APIServerSecureRouting](#feature-gate-1-apiserversecurerouting)
    - [Feature Gate 2: APIServerLegacyPortDisable](#feature-gate-2-apiserverlegacyportdisable)
    - [Configuration Example](#configuration-example)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)

## Summary
This proposal introduces secure routing mechanisms for Gardener's API server proxy to prevent security vulnerabilities when using opaque LoadBalancers.

## Motivation

> The current architecture of Gardener, when used with opaque/instransparent LoadBalancer configurations and the ACL extension, can result in a security vulnerability. This vulnerability stems from the addition of multiple proxy protocol headers to network packets, which can lead to possible unauthorized access attempts to incorrect kube-api servers.

Reviewer: ```suggestion
The current architecture of Gardener, when used with opaque/intransparent LoadBalancer configurations and the ACL extension, can result in a security vulnerability. This vulnerability stems from the addition of multiple proxy protocol headers to network packets, which can lead to possible unauthorized access attempts to incorrect kube-api servers.
```

> Alternatively, reducing the number of supported connection architectures by removing the proxy protocol method would simplify the system. The remaining two architectures already utilize HTTP CONNECT and are inherently immune to these security concerns, making them more reliable choices for secure communication.

Reviewer: Please do not call different protocols "connection architectures". Maybe we can replace "connection architectures" with "connection types". Alternatively, protocol would also be possible.

The second sentence should be reworked. The SNI port uses TLS (with server name indication). The existing VPN port uses (unencrypted) HTTP Connect with TLS-style payload. Hence, the ports use rather different protocols, but they may both be immune to the special vulnerability of using multiple proxy protocol headers.

> The proposed solution involves implementing a custom header for secure routing using HTTP CONNECT (as currently used by the VPN listener), introducing a new port or path for API traffic, and gradually transitioning to this system through a feature gate.

Reviewer: This makes it sound like we want to introduce a new port. I thought from previous discussions that we want to use the VPN port for this, which I prefer.
Please clarify.

### Goals
- Eliminate the vulnerability caused by dual proxy protocol headers
- Ensure secure routing of traffic to the correct API server
- Provide a clear migration path for existing Gardener deployments

> - Allow [`gardener-extension-acl`](https://github.com/stackitcloud/gardener-extension-acl) to be used with proxy protocol on (non-transparent) LoadBalancers

Reviewer: Do we want to add the reuse of the VPN infrastructure as a goal? I would like to have a single port responsible for handling HTTP Connect requests be it for VPN or apiserver-proxy.
WDYT?

Reviewer 2: Sounds reasonable to me

### Non-Goals
- Modifying the core functionality of the ACL extension

## Proposal

The proposed solution involves the following key changes:

- Reconfigure the `istio-ingressgateway` to use the new routing method

>     - Introduce a new port/path for API traffic from apiserver-proxy

Reviewer: ```suggestion
    - Introduce a new path for API traffic from apiserver-proxy reusing the VPN infrastructure
```

    - Implement a new custom header for secure API server routing
- Reconfigure the `apiserver-proxy` to use HTTP CONNECT for secure tunneling
- Develop a feature gate to control the rollout of the new routing mechanism
- Provide a phased implementation approach for gradual adoption

## Concerns with the [Current Architecture](./11-apiserver-network-proxy.md)

> The `apiserver-proxy` creates a proxy protocol header for each connection, containing information about the source and destination of the traffic. This header gets forwarded by the LoadBalancer towards the `istio-ingressgateway`.

Reviewer: Maybe we should also link to the proxy protocol specification, e.g. https://www.haproxy.org/download/3.2/doc/proxy-protocol.txt. Unfortunately, the version in the URL is the HA proxy version...

A second proxy protocol header may be created when:
- The LoadBalancer is configured through the [ACL extension](https://github.com/stackitcloud/gardener-extension-acl)
- The LoadBalancer is operating in an opaque/intransparent mode

> - Proxy protocol is explicitly enabled in the LoadBalancer configuration

Reviewer: From my point of view the order is weird and it is not clear whether all of those options need to apply or only a single one. I propose the following, but feel free to rework as you desire.

```suggestion
A second proxy protocol header may be created when proxy protocol is explicitly enabled in the LoadBalancer configuration. Usually, this is done to preserve the client IP address in case of an opaque/intransparent LoadBalancer. Without client IP address preservation certain operations based on the client IP address are not possible. A prime example requiring client IP address preservation is IP allowlisting with the [ACL extension](https://github.com/stackitcloud/gardener-extension-acl).
```

In cases where dual proxy protocol headers exist, we require specific information from each header:
- The destination IP address from the `apiserver-proxy`'s proxy protocol header
- The source IP address from the LoadBalancer's proxy protocol header

However, the `istio-ingressgateway` processes these headers sequentially, reading the first proxy protocol header and then overwriting its information with the second proxy protocol header. This behavior means only the information from the second header is ultimately used for traffic filtering decisions in the [ACL extension](https://github.com/stackitcloud/gardener-extension-acl).

This situation becomes problematic under these conditions:

> - When [ACL extension](https://github.com/stackitcloud/gardener-extension-acl) is actively filtering traffic based on source IP addresses

Reviewer: ```suggestion
- When the [ACL extension](https://github.com/stackitcloud/gardener-extension-acl) is actively filtering traffic based on source IP addresses
```

> - When multiple shoot clusters are using overlapping private IP ranges

Reviewer: I do not get why the overlapping private IP ranges have any relevance here.
Could you please elaborate?

Reviewer 2: I don't think overlapping private IP ranges have a relevance here. This seems to be a misunderstanding from an internal discussion.

- When the LoadBalancer is not maintaining the original client source IP information

The issue does not impact environments where:
- The LoadBalancer operates in transparent mode
- Only a single proxy protocol header is present

> - Alternative authentication methods are the primary security control

Reviewer: As far as I know, Gardener always configures `kube-apiserver` to not accept anonymously requests. Therefore, there is always an additional authentication already in place.
This makes me wonder whether this point makes sense. Did you have something else in mind or am I somehow missing the point?

Reviewer 2: I think this was supposed to say that even if the ACL-based authorization is circumvented, a client still doesn't gain authorized access to the API server, as it needs to present valid credentials first.

> - Client IP ranges are strictly segregated between shoots

Reviewer: See above. I do not get how segregated IP ranges make a difference here. Please clarify.

![Current State with opaque LB](./assets/30-current-architecture.png "Current State with opaque LB")

## Proposed Changes

1. Custom Header Implementation
    - Introduce a new custom header `X-Gardener-API-Route`
    - Define the structure and encoding of the header value to include necessary routing information

2. New Port/Path
    - Implement a new port or path on the `istio-ingressgateway` for the new routing mechanism

>     - Ensure this new route is separate from any existing VPN or tunnel infrastructure

Reviewer: As mentioned before, I prefer to reuse the existing port.

```suggestion
2. New Path
    - Implement a new path on the `istio-ingressgateway` for the new routing mechanism
    - Ensure this new route is separate from any existing VPN, but reuses the same tunnel infrastructure
```

3. HTTP CONNECT Reconfiguration:

>     - Reconfigure the `apiserver-proxy` to use HTTP CONNECT for establishing a secure tunnel

Reviewer: Nit: Not sure how this tunnel is "secure". The existing architecture is based on the fact that the payload is already TLS-encrypted. Hence, further encryption is not required. However, the VPN tunnel using pure HTTP Connect is not really secure in itself as there is no encryption happening.

Reviewer 2: Right. I will remove all mentions of the term "secure tunnel" for HTTP proxying in the next iteration of the document.

    - This will replace the current TCP proxy protocol method

>     - The HTTP CONNECT method will be used on the newly implemented port/path

Reviewer: ```suggestion
    - The HTTP CONNECT method will be used on the newly implemented path
```

4. Feature Gate
    - Implement a feature gate to control the rollout of the new routing mechanism
    - Define stages for the feature gate: alpha, beta, and stable

5. Phased Implementation

>     - Phase 1: Add the new port/path and custom header processing

Reviewer: ```suggestion
    - Phase 1: Add the new path and custom header processing
```

>     - Phase 2: Gradually reconfigure shoots to use the new routing method

Reviewer: I assume that you mean to enable the feature gate per seed or do you have something else in mind how to gradually roll this change out?

```suggestion
    - Phase 2: Gradually reconfigure shoots to use the new routing method by enabling the feature gate per seed
```

Reviewer 2: We once discussed internally to allow configuring the feature gate per shoot. This has been done before for other feature gates in gardener but I agree that this shouldn't be done in this case.

    - Phase 3: Deprecate and remove the old routing method

![Proposed Architecture](./assets/30-proposed-architecture-http-connect.png "Proposed Architecture")

> ![Proposed Architecture](./assets/30-proposed-architecture-tls.png "Proposed Architecture")

Reviewer: I am slightly confused by the diagrams. What should they highlight?

The HTTP Connect picture seem to be the target state, but what does the TLS picture mean? If it should show the second path from pod/workload to `kube-apiserver` it is simply wrong as the pure TLS/SNI path will not go through the `apiserver-proxy`. However, there is also no plan as far as I understand this GEP to do TLS from `apiserver-proxy` to `kube-apiserver`.

Please clarify.

Reviewer 2: I'm also confused by the diagrams. I will remove the second one with the next iteration of the document and revisit the first one.

    
    
## Custom Header Specification

The new custom header will be structured as follows:

```
X-Gardener-API-Route: <routing-information>
```

The routing information will be a concatenated string, joined by pipes, containing the following fields:

- `direction`: The direction of traffic.
- `port`: The destination port.
- `subset`: Istio subset/version.
- `destination`: The destination.

Putting this to practice should result in a header formatted like this:

> `X-Gardener-API-Route: outbound|8134||kube-apiserver.foobar.svc.cluster.local`

Reviewer: As you might have guessed by now, the port triggers me again. I prefer the following:

```suggestion
`X-Gardener-API-Route: outbound|8132||kube-apiserver.foobar.svc.cluster.local`
```

Reviewer 2: The VPN client uses `outbound|1194||vpn-seed-server-*.shoot--foo--bar.svc.cluster.local` and the API server proxy will use `outbound|443||kube-apiserver.shoot--foo--bar.svc.cluster.local`.
`8132` is the istio-ingressgateway port, which is not related to the upstream cluster specified in the custom header.

During the transition period, the system should be able to handle both the new custom header and the existing routing method. This can be controlled via the feature gate.

If the header is missing, malformed, or fails verification, the error should be logged with appropriate details for debugging and return an HTTP 400 (Bad Request) status code. It should never fall back to the old routing method to ensure security.

> ## HTTP CONNECT Implementation

Reviewer: Let's also link for HTTP Connect, e.g. https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.6.

The `apiserver-proxy` will be reconfigured to use HTTP CONNECT for establishing a secure tunnel to the `kube-apiserver`. This involves:

> 1. Initiating an HTTP CONNECT request to the new port/path on the `istio-ingressgateway`

Reviewer: ```suggestion
1. Initiating an HTTP CONNECT request to the new path on the `istio-ingressgateway`
```

2. Including the new custom header in the CONNECT request
3. Creating a secure connection over the established tunnel once the CONNECT request is accepted

> 4. Forwarding the original API server request through this secure tunnel

Reviewer: Nit: As mentioned above, calling this tunnel secure is a bit of a stretch.

```yaml
...
- domains: 
  - api.*
  name: gardener-api-route
  routes:
  - match:
      connect_matcher: {}
    route:
     cluster_header: X-Gardener-API-Route
     upgrade_configs:
     - connect_config: {}
       upgrade_type: CONNECT
...
```

This solution would be reusing existing Envoy filtering on the `apiserver-proxy` pod and simply exchange its Proxy Protocol configuration for the new proposed header.

### Istio IngressGateway Configuration

Similarly, it would be required to configure the Istio IngressGateway to accept HTTP CONNECT requests and route them appropriately. This could e.g. be achieved utilizing Istio Custom Resources (Gateway, VirtualService):

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: api-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
    ...
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: api-routes
  namespace: istio-system
spec:
  ...
  gateways:
  - api-gateway
  http:
  - match:
    - method:
        exact: CONNECT
      headers:
        X-Gardener-API-Route:

>           regex: ".*"

Reviewer: We probably do not want to do this, but instead use an external authorization service just like with the VPN. It should not be possible to reach random kubernetes services in the seed using this approach. After all, we want to close security vulnerabilities and not open new ones ;-) .
Maybe you just link this to the next section and make it explicit that not just any random header value will be accepted.

    ...
```

This configuration sets up the Istio IngressGateway to accept HTTPS traffic and route HTTP CONNECT requests with the proposed custom header to the `kube-apiserver`.

### EnvoyFilter for Custom Header Processing

To process the proposed custom header and make routing decisions based on it, it will be required to additionally add and/or modify an EnvoyFilter. Again, existing configuration may be reused adjusted to this new use case to keep convention parity within the code base.

```yaml
{{- if eq .Values.vpn.enabled true -}}
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: reversed-vpn-auth-server
  namespace: {{ .Release.Namespace }}
spec:
  selector:
{{ .Values.labels | toYaml | indent 4 }}
  servers:
  - hosts:
    - reversed-vpn-auth-server.garden.svc.cluster.local
    port:
      name: tls-tunnel

>       number: 8134

Reviewer: ```suggestion
      number: 8132
```

      protocol: HTTP
{{- end }}
```
### Implementation Steps

1. Apply the Istio Gateway and VirtualService configurations.
2. Deploy the EnvoyFilter for custom header processing.
3. Update any relevant Istio authorization policies to allow the new traffic flow.
4. Update the `apiserver-proxy` code or configuration to use HTTP CONNECT and include the custom header.
5. Deploy the updated `apiserver-proxy` configuration.

## Phased Rollout

The implementation will be controlled by two distinct feature gates that handle different aspects of the solution. Each feature gate can be disabled per shoot to allow testing of the old implementation via E2E tests.

> ### Feature Gate 1: `APIServerSecureRouting`

Reviewer: Let's not confuse people to think there is some general routing problem with API servers.

```suggestion
### Feature Gate 1: `APIServerProxySecureRouting`
```

Controls whether shoots use the new secure routing implementation with HTTP CONNECT and custom headers.

**Rollout Plan:**
- Initially introduced as disabled by default
- Once stability is proven in production environments, enabled by default for new shoots
- Existing shoots will retain their previous setting until explicitly migrated

> - When fully proven, the feature gate will be removed and the functionality will become permanent

Reviewer: I get the intention by this approach, but I do not understand how you want to achieve this. Could you please clarify? In other words: How can the reconciliation figure out if this is a new cluster or not?

This feels a bit weird to me. I would have assumed that all shoot clusters hosted on a given seed will automatically migrate to the new approach once the feature gate is enabled.

> ### Feature Gate 2: `APIServerLegacyPortDisable`

Reviewer: Same as above:

```suggestion
### Feature Gate 2: `APIServerProxyLegacyPortDisable`
```

> Controls whether the legacy port is available. This feature gate can only be enabled after `APIServerSecureRouting` has been fully rolled out.

Reviewer: ```suggestion
Controls whether the legacy port is available. This feature gate can only be enabled after `APIServerProxySecureRouting` has been fully rolled out.
```

**Rollout Plan:**
- Introduced as disabled by default
- Once all shoots have migrated to secure routing, enabled by default for new shoots
- Existing shoots will be notified to migrate during reconciliation

> - When migration is complete, the feature gate will be removed and the legacy port will be permanently disabled

Reviewer: I do not get why we need this feature gate at all. Also, I see no reason to notify anybody for a migration. This should be done automatically from my point of view.

> ---

Reviewer: Please add a link to the GEP to https://github.com/gardener/gardener/blob/master/docs/README.md.

status: implementable
authors:
- "@robinschneider"
- "@knht"
reviewers:
- "@timebertt"
---

# GEP-30: Rework apiserver-proxy to drop PROXY protocol

## Table of Contents

- [Summary](#summary)
- [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
- [Proposal](#proposal)
- [Current Architecture](#current-architecture)
- [Proposed Changes](#proposed-changes)
- [Custom Header Specification](#custom-header-specification)
- [HTTP CONNECT Implementation](#http-connect-implementation)
    - [Technical Implementation Details](#technical-implementation-details)
    - [Istio IngressGateway Configuration](#istio-ingressgateway-configuration)
    - [EnvoyFilter for Custom Header Processing](#envoyfilter-for-custom-header-processing)
    - [Implementation Steps](#implementation-steps)
- [Feature Gate Implementation](#feature-gate-implementation)
    - [Feature Gate 1: APIServerSecureRouting](#feature-gate-1-apiserversecurerouting)
    - [Feature Gate 2: APIServerLegacyPortDisable](#feature-gate-2-apiserverlegacyportdisable)
    - [Configuration Example](#configuration-example)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)

## Summary

> This proposal addresses a critical security vulnerability in the Gardener's API server proxy configuration when used with opaque or intransparent LoadBalancers, particularly in conjunction with the ACL extension. The issue involves the potential for dual proxy protocol headers, which can lead to misrouting, information leakage, and possible unauthorized access attempts. The proposed solution is to implement a custom header for secure routing using HTTP CONNECT (as currently used by the VPN listener), introduce a new port or path for API traffic, and gradually transition to this system through a feature gate.

Reviewer: From my point of view, most of what is stated here is the motivation and not the summary. I would have expected the short summary to be "replace proxy protocol with http connect", but this paragraph contains a lot more.

Please reconsider what you add to the summary section and what would be better suited in another section.

## Motivation

> The current architecture of Gardener, when used with opaque/instransparent LoadBalancer configurations and the ACL extension, can result in a security vulnerability. This vulnerability stems from the addition of multiple proxy protocol headers to network packets, which can lead to:

Reviewer: Please add links to the ACL extension.

1. Misrouting of traffic
2. Potential information leakage

> 3. Possible unauthorized access attempts to incorrect kube-api servers

Reviewer: The order does not seem right. It is also not obvious why this would be a "critical security vulnerability". 1. is a functional issue. 2. looks like a potential security issue. 3. is kind of an overstatement as it not possible to circumvent authorisation in `kube-apiserver`, but it might be possible to get around IP allow-listing.

Overall, I would have hoped for a bit more details in the motivation section to understand why we should switch.

### Goals
- Eliminate the vulnerability caused by dual proxy protocol headers
- Ensure secure routing of traffic to the correct API server
- Provide a clear migration path for existing Gardener deployments

> - Maintain or improve the overall performance and reliability of the system

Reviewer: Performance/reliability improvements are a somewhat surprising goal as part of this GEP. I am keen to find out what you mean by that.

Author: The goal of the last point wasn't a direct improvement as part of the GEP, but rather maintaining the current overall performance to not regress in that regard. I'll update the relevant section to remove the improvement part.

> - Allow `gardener-extension-acl` to be used with proxy protocol on (non-transparent) LoadBalancers

Reviewer: Link to the extension please.

### Non-Goals
- Modifying the core functionality of the ACL extension

> - Addressing security issues unrelated to the proxy protocol headers and API server routing

Reviewer: Kind of strange statement. Are you then targeting everything else, e.g. world domination? From my perspective, it is fairly obvious that the GEP has on target and does not try to change random things. Therefore, I find this line a bit surprising.

Author: World domination is planned for GEP-31 :-)

Reviewer: Maybe I should take a closer look at https://github.com/gardener/gardener/pull/10828 then ;-) ...

> - Reusing existing VPN or tunnel-related infrastructure

Reviewer: Not sure I understand this. Why do you want to replicate things instead of re-using shared infrastructure? It seems sensible to have a common port for http connect instead of replicating it to two.

Author: At the end of the day we'd like to reuse the same istio-ingressgateway, the other parts of the system, etc.. What we don't want to reuse is the actual Header in place right now. This needs to be written more carefully and precise in our next iteration, thanks.

## Proposal

The proposed solution involves the following key changes:

- Reconfigure the `istio-ingressgateway` to use the new routing method
    - Introduce a new port/path for API traffic from apiserver-proxy
    - Implement a new custom header for secure API server routing
- Reconfigure the `apiserver-proxy` to use HTTP CONNECT for secure tunneling
- Develop a feature gate to control the rollout of the new routing mechanism
- Provide a phased implementation approach for gradual adoption

> ## Current Architecture

Reviewer: Please add at least a link to GEP-11 as it contains the concept of the current architecture.

With the current architecture, the `apiserver-proxy` creates a proxy protocol header, which gets forwarded by the LoadBalancer towards the `istio-ingressgateway`.

> A second proxy protocol header is created by the LoadBalancer, which is configured through the ACL extension.

Reviewer: Please describe this in more detail stating general facts and options, e.g. "A second proxy protocol header may be created if ...".

There are boundary conditions when this is relevant and there are situations when it does not matter. Please add this context.

> At this point, we need the destination IP address from the `apiserver-proxy` proxy protocol header and the source IP from the LoadBalancer proxy protocol header.

Reviewer: Please rephrase and clarify this sentence.

Unfortunately, the `istio-ingressgateway` will read the first proxy protocol header and then overwrites the information with the second proxy protocol header and only uses these source and destination IP addresses.

> These source IP addresses will be used for filtering allowed traffic.

Reviewer: ... in the acl extension.

> Instead of the public IP address from the router *(here 10.1.0.1)* it will allow traffic from the client-pod *(here 10.3.0.1)*. This causes other clients in other shoots with the same IP address to bypass the list of allowed clients by the ACL extension and cause unauthorized requests to the `kube-apiserver`.

Reviewer: Please change this paragraph as it is incorrect and misleading.

A router is a specific component in networking, but I do not see the term used correctly here. There is no bypass of IP allow list checks and there is also no cross-shoot problem.

![Current State with opaque LB](./assets/30-current-architecture.png "Current State with opaque LB")

## Proposed Changes

1. Custom Header Implementation

>     - Introduce a new custom header, e.g. `X-Gardener-API-Route`

Reviewer: What do you mean with "for example"? Should this not be part of this enhancement proposal?

Do you have a plan or only a concept of a plan?

    - Define the structure and encoding of the header value to include necessary routing information

2. New Port/Path
    - Implement a new port or path on the `istio-ingressgateway` for the new routing mechanism
    - Ensure this new route is separate from any existing VPN or tunnel infrastructure

3. HTTP CONNECT Reconfiguration:
    - Reconfigure the `apiserver-proxy` to use HTTP CONNECT for establishing a secure tunnel
    - This will replace the current TCP proxy protocol method
    - The HTTP CONNECT method will be used on the newly implemented port/path

4. Feature Gate
    - Implement a feature gate to control the rollout of the new routing mechanism
    - Define stages for the feature gate: alpha, beta, and stable

5. Phased Implementation
    - Phase 1: Add the new port/path and custom header processing
    - Phase 2: Gradually reconfigure shoots to use the new routing method
    - Phase 3: Deprecate and remove the old routing method

6. Component Reconfigurations:
    - Modify `istio-ingressgateway` to process the new header and route traffic accordingly

>     - Update `apiserver-proxy` to add the new custom header

Reviewer: This surprised me a bit.
The modification of istio seems to fit to 2.
The change in `apiserver-proxy` seems to be included in 3.

What am I missing?

![Proposed Architecture](./assets/30-proposed-architecture.png "Proposed Architecture")
    
    
## Custom Header Specification

The new custom header will be structured as follows:

```
X-Gardener-API-Route: <routing-information>
```

The routing information will be a concatenated string, joined by pipes, containing the following fields:

```json
{
  "direction": "outbound",
  "port": "8134",
  "subset": null,
  "destination": "kube-apiserver.{{ .namespace }}.svc.cluster.local"
}

> ```

Reviewer: Why print a `json` structure if the header is not `json`?

- `direction`: The direction of traffic.
- `port`: The destination port.
- `subset`: Istio subset/version.
- `destination`: The destination.

Putting this to practice should result in a header formatted like this:

> `X-Gardener-API-Route: outbound|8134||kube-apiserver.{{ .namespace }}.svc.cluster.local`

Reviewer: `{{ .namespace }}` suggests that go templating. Please rather use an example instead here.

Or formatted in YAML in a potential deployment manifest

```yaml
- name: X-Gardener-API-Route
  value: outbound|8134||kube-apiserver.{{ .namespace }}.svc.cluster.local

> ```

Reviewer: Personally, I find this redundancy not useful. Please rather remove it to keep the document concise.

During the transition period, the system should be able to handle both the new custom header and the existing routing method. This can be controlled via the feature gate.

If the header is missing, malformed, or fails verification, the error should be logged with appropriate details for debugging and return an HTTP 400 (Bad Request) status code. It should never fall back to the old routing method to ensure security.

## HTTP CONNECT Implementation

The `apiserver-proxy` will be reconfigured to use HTTP CONNECT for establishing a secure tunnel to the `kube-apiserver`. This involves:

1. Initiating an HTTP CONNECT request to the new port/path on the `istio-ingressgateway`
2. Including the new custom header in the CONNECT request

> 3. Establishing a secure tunnel once the CONNECT request is accepted

Reviewer: What do you mean with "secure tunnel" in this context? Could you please elaborate?

### Technical Implementation Details

The `apiserver-proxy` will be required to be updated to use HTTP CONNECT. This should involve modifying its configuration and potentially its code. One example of how the configuration might look like after reusing its Envoy tunneling config:

```yaml
...
- domains: 
  - api.*
  name: gardener-api-route
  routes:
  - match:
      connect_matcher: {}
    route:
     cluster_header: X-Gardener-API-Route
     upgrade_configs:
     - connect_config: {}
       upgrade_type: CONNECT
...
```

This solution would be reusing existing Envoy filtering on the `apiserver-proxy` pod and simply exchange its Proxy Protocol configuration for the new proposed header.

### Istio IngressGateway Configuration

Similarly, it would be required to configure the Istio IngressGateway to accept HTTP CONNECT requests and route them appropriately. This could e.g. be achieved utilizing Istio Custom Resources (Gateway, VirtualService):

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: api-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
    ...

>     ...

Reviewer: ```suggestion
```

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: api-routes
  namespace: istio-system
spec:
  ...

>   ...

Reviewer: ```suggestion
```

  gateways:
  - api-gateway
  http:
  - match:
    - method:
        exact: CONNECT
      headers:
        X-Gardener-API-Route:
          regex: ".*"
    ...

>     ...

Reviewer: ```suggestion
```

```

This configuration sets up the Istio IngressGateway to accept HTTPS traffic and route HTTP CONNECT requests with the proposed custom header to the `kube-apiserver`.

### EnvoyFilter for Custom Header Processing

To process the proposed custom header and make routing decisions based on it, it will be required to additionally add and/or modify an EnvoyFilter. Again, existing configuration may be reused adjusted to this new use case to keep convention parity within the code base.

```yaml
{{- if eq .Values.vpn.enabled true -}}
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: reversed-vpn-auth-server
  namespace: {{ .Release.Namespace }}
spec:
  selector:
{{ .Values.labels | toYaml | indent 4 }}
  servers:
  - hosts:
    - reversed-vpn-auth-server.garden.svc.cluster.local
    port:
      name: tls-tunnel
      number: 8134
      protocol: HTTP
{{- end }}

> ```

Reviewer: This seems to contradict the following non-goal:
`- Reusing existing VPN or tunnel-related infrastructure`

Frankly, I would like to understand why we should add another port for the new traffic path instead of consolidating the existing ones. In other words: Why should it not be the goal to use the same traffic path as the VPN and thereby mid-term reduce the amount of used ports?

Author: Thanks for this recommendation, we haven't considered this and will change appropriately.

### Implementation Steps

1. Apply the Istio Gateway and VirtualService configurations.
2. Deploy the EnvoyFilter for custom header processing.
3. Update any relevant Istio authorization policies to allow the new traffic flow.
4. Update the `apiserver-proxy` code or configuration to use HTTP CONNECT and include the custom header.
5. Deploy the updated `apiserver-proxy` configuration.

## Phased Rollout

The implementation will be controlled by two distinct feature gates that handle different aspects of the solution. Each feature gate can be disabled per shoot to allow testing of the old implementation via E2E tests.

### Feature Gate 1: `APIServerSecureRouting`
Controls the enablement of the new routing implementation using HTTP CONNECT and custom headers.

#### Alpha
- **Default:** Disabled
- **Actions:**
    - Introduces new port for HTTP CONNECT
    - Deploys new custom header processing

>     - Configures mTLS for new routing path

Reviewer: Where is mTLS now coming from and why would we need it?

    - Enables E2E testing of the feature
    - Implementation of all unit and integration tests

>     - Documentation completion

Reviewer: I do not like this documentation. Obviously, first things need to be implemented, tested and documented, but this has nothing to do with the state of the feature gate. Why do you make it look like there is a direct correlation with the feature gate status?

#### Beta
- **Default:** Enabled
- New shoots automatically use new implementation
- Existing shoots continue using old implementation

#### GA
- **Default:** Always enabled
- All shoots use new implementation

> - Existing shoots automatically migrated during reconciliation

Reviewer: This would be a very non-standard way of dealing with feature gates. Usually, they are binary, i.e. on or off, but here you actually propose a trinary state where GA is different from beta turned on. Please reconsider.

### Feature Gate 2: `APIServerLegacyPortDisable`
Controls the disabling of the legacy port. Can only progress once `APIServerSecureRouting` is GA.

#### Alpha
- **Default:** Disabled
- **Prerequisites:**
    - Feature gate `APIServerSecureRouting` is in GA state
    - All monitored shoots successfully using the new implementation
- **Actions:**
    - Marks port as deprecated
    - Implementation of E2E tests with legacy port disabled

#### Beta
- **Default:** Enabled
- Automatically closes legacy port for new shoots
- Existing shoots prompted for migration during reconciliation

#### GA
- **Default:** Always enabled
- Legacy port completely disabled

> - All configuration for legacy port removed

Reviewer: I do not understand why we would need a feature gate for this. Once all shoots use the new implementation, which is stated as a prerequisite, there is no need for the old port anyway. Why not just close it? This seems to be a lot of overhead/wasted time for nothing.

Author: Our idea was to introduce this second feature gate to control if previous rollouts of the first feature gate had happened yet. This would more or less allow for little to no business logic in Gardener itself but rather allow for manual configuration once every shoot has been reconciled.

### Testing Strategy

> - Feature gates can be switched off per shoot to allow testing of the old implementation

Reviewer: I assumed that the feature gate would be on the `gardenlet` level, but you actually never really mention it. Could you please clarify?

In case the feature gate is on `gardenlet` level how do you envision to disable the feature gate per shoot cluster? I do not get this part.

- E2E tests will cover both new and old implementations
- All testing functionality is implemented and validated in alpha phase

## Alternatives

> - Implement stricter filtering of proxy protocol headers at the istio-ingress gateway level

Reviewer: Could you please elaborate? I would be very interested in how this problem can be solved like this.

Author: This can be resolved as well as there is no way of implementing custom filtering on L3/4 in Envoy without maintaining a custom fork.

> - Use a custom LoadBalancer solution that is transparent or aware of the Gardener architecture

Reviewer: Not all infrastructures provide transparent load balancers. Requiring a transparent load balancer for Gardener is no solution. However, it is fair to have this as a prerequisite for the ACL extension.

In all seriousness, we are not aware of any other or real alternative solution to address this issue.

> ---

Reviewer: As mentioned before, I am missing how this is related to the existing traffic paths in istio, especially the VPN path. What is the masterplan if there is one? I would like to see a consolidation of traffic paths, but I do not like to add additional ports on the istio. Please clarify.

Author: Clarified in previous comment.

  - [Visualization of the Proposed Architecture](#visualization-of-the-proposed-architecture)

## Summary

This proposal reworks the API server proxy (originally introduced in [GEP-08](08-shoot-apiserver-via-sni.md)) to use [HTTP CONNECT requests](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.6) (i.e., HTTP proxy) instead of the [proxy protocol](https://www.haproxy.org/download/3.2/doc/proxy-protocol.txt) for connecting in-cluster clients on the shoot side to the corresponding API server on the seed side.
With this, the API server proxy uses the same network infrastructure and protocol to connect to the shoot control plane as the shoot's VPN client (see [GEP-14](14-reversed-cluster-vpn.md)).

The rework allows more scenarios like using the [ACL extension](https://github.com/stackitcloud/gardener-extension-acl) with opaque (non-transparent) LoadBalancers on the seed side that rely on the proxy protocol themselves to preserve the client's IP.

## Motivation

Since [GEP-08](08-shoot-apiserver-via-sni.md) introduced shared LoadBalancers for shoot control planes on the seed side, clients need to indicate which control plane they want to connect to through the LoadBalancer.

> The Envoy proxy in the Istio ingress gateway receives the traffic from the shared LoadBalancer and is responsible routing traffic to the indicated control plane.

Reviewer: ```suggestion
The Envoy proxy in the Istio ingress gateway receives the traffic from the shared LoadBalancer and is responsible for routing traffic to the indicated control plane.
```

For this, Gardener currently uses different protocols based on the connection type:

> - When connecting to a shoot's API server directly, this is done using TLS SNI (Server Name Indication). I.e., the destination API server is indicated by the hostname in the SNI header of the TLS handshake.

Reviewer: May just be nitpicking, but the term headers is usually used in conjunction with http. In TLS, SNI is an extension in the TLS client hello packet.

```suggestion
- When connecting to a shoot's API server directly, this is done using TLS SNI (Server Name Indication). I.e., the destination API server is indicated by the hostname in the SNI extension of the TLS client hello packet.
```

> - When connecting to a shoot's API server via the `kubernetes` Service (fallback to the previous protocol for in-cluster clients), the SNI header is set to the same value (`kubernetes.default.svc.cluster.local`) on all shoots and cannot be used to indicate the destination API server. Therefore, the API server proxy handles traffic on this service and prepends a proxy protocol header with a shoot-specific destination IP to indicate the destination.

Reviewer: What do you mean with `(fallback to the previous protocol for in-cluster clients)`?

As far as I know, it is a standard kubernetes API to have a service called `kubernetes` in the `default` namespace representing an endpoint of the cluster's API server. It is only available for the in-cluster clients, but I do not see this as a fallback, but rather as an alternative option. Using the `kubernetes` service has the advantage of being cluster-independent, i.e. can be used in all kubernetes clusters, while using the "real" domain name is very cluster-dependent.

In short, I am a bit confused by the wording in the parentheses. Please either remove it completely or rephrase it.

Reviewer 2: What I had in mind was the following:
In-cluster clients typically read the `KUBERNETES_SERVICE_HOST` env var (at least client-go does so) and connect to this host directly which would be the internal API server domain (as injected into all pods by gardener) instead of `kubernetes.default.svc`.
Hence, I would consider `kubernetes.default.svc` and the API server proxy a fallback path for clients which do not respect the `KUBERNETES_SERVICE_HOST` env var.

I get why the wording was confusing. I removed it because it's not really relevant for describing the different connection types in gardener in the scope of this GEP.

- When connecting to a shoot's VPN server, the shoot VPN client sends an HTTP CONNECT request to the shared LoadBalancer and indicates the destination by adding the `Reversed-VPN` HTTP header with the Envoy cluster string as a value (e.g., `outbound|1194||vpn-seed-server-0.shoot--foo--bar.svc.cluster.local`). I.e., it uses the ingress gateway as an HTTP proxy. In contrast to usual HTTP proxies, the target in the CONNECT request line is discarded.

Note that in all cases the payload (HTTP request or OpenVPN tunnel) is end-to-end encrypted even if it is tunneled via an unencrypted HTTP connection.

> Shoot owners can use the [ACL extension](https://github.com/stackitcloud/gardener-extension-acl) for restricting traffic to the control plane based on client IPs – on all three of the described connection types.
> In seed setups where only opaque LoadBalancers are available, the Gardener operator needs to configure the LoadBalancer to use the proxy protocol to preserve the original client IP.
> With the proxy protocol, the original client IP is lost and the ACL extension cannot restrict the traffic as configured.

Reviewer: ```suggestion
Shoot owners can use the [ACL extension](https://github.com/stackitcloud/gardener-extension-acl) for restricting traffic to the control plane based on client IPs – on all three of the described connection types.
In seed setups where only opaque LoadBalancers are available, the Gardener operator needs to configure the LoadBalancer to use the proxy protocol to preserve the original client IP.
Without the proxy protocol, the original client IP is lost and the ACL extension cannot restrict the traffic as configured.
```

However, the value of the `X-Gardener-Destination` header follows the same format (i.e., the Envoy cluster string format):

```text
outbound|443||kube-apiserver.<technical-shoot-id>.svc.cluster.local
```

In the corresponding upstream cluster section of the config, the destination port to connect to on istio-ingressgateway is changed from 8443 (`proxy`) to 8132 (`tls-tunnel`).
This configures the API server proxy to reuse the existing VPN infrastructure on the seed side.
Finally, the `transport_socket` is removed which disables adding the proxy protocol headers.

With this, a connection is established as follows:

1. An in-cluster client (e.g., pod) opens a TLS connection to `kubernetes.default.svc.cluster.local`.
2. The Envoy process in the API server proxy pod on the same node sends an HTTP CONNECT request to the API server domain of the shoot, i.e., to the ingress gateway on the corresponding seed.
3. The ingress gateway discards the target from the HTTP request line and opens a TCP connection to the upstream cluster indicated by the `X-Gardener-Destination` header, i.e., in-cluster service of the shoot cluster.

> 4. The TLS payload is proxied from the in-cluster client via the API server proxy and the ingress gateway to the shoot API server.

Reviewer: From my point of view, the logical connection between 1 and 2 is missing. It can either be added to the first point that the domain name points to an address in use by apiserver-proxy/envoy or you can add a new point noting this fact.
WDYT?

                   port_redirect: 443
-                typed_per_filter_config:
-                  envoy.filters.http.ext_authz:
-                    '@type': type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthzPerRoute
-                    disabled: true

# ...

-  - applyTo: HTTP_FILTER
-    match:
-      context: GATEWAY
-      listener:
-        filterChain:
-          filter:
-            name: envoy.filters.network.http_connection_manager
-            subFilter:
-              name: envoy.filters.http.router
-        portNumber: 8132
-    patch:
-      filterClass: AUTHZ
-      operation: INSERT_BEFORE
-      value:
-        name: envoy.filters.http.ext_authz
-        typed_config:
-          '@type': type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
-          grpc_service:
-            envoy_grpc:
-              cluster_name: outbound|9001||reversed-vpn-auth-server.garden.svc.cluster.local
```

The existing route matcher in the `reversed-vpn` virtual host is extended to additionally check the `Reversed-VPN` header.
The header's value must match the given regex, so that only VPN servers can be specified as the proxy destination.

> Then, a new route is added to the same virtual host that matches and validates the new `X-Gardener-Destination` header but also allows shoot API servers as valid destinations.

Reviewer: It is probably worth adding why the `X-Gardener-Destination` header should support the VPN connection as well. Without reasoning it looks like this is unnecessary, but if we want to eventually unify this, i.e. remove `Reversed-VPN`, it makes a lot of sense.
I propose to add a short sentence why we want this. Maybe a link to the next section is also sufficient.
WDYT?

Reviewer 2: Good point, added a sentence and link.

   - Remove the `UsesUnifiedHTTPProxyPort` constraint in the `Shoot` status.

After these steps, the legacy VPN network infrastructure on the seed side (including the `tls-tunnel` port 8132) is removed. Instead, the unifying infrastructure using the `http-proxy` port 8443 is used.

In this plan, we need the feature gates for guarding the removal of one of the ingress gateway ports because we cannot be sure that a given Gardener version has been able to successfully reconfigure the related components in all shoot clusters.
I.e., we cannot assume that the port in question is no longer used, because the shoot reconciliation flow might not have reached the relevant tasks.
This might be because the shoot is marked to be ignored or an `Extension` with lifecycle `BeforeKubeAPIServer` cannot be reconciled successfully.
The feature gate allows the Gardener operator to delay the removal of the ingress gateway port if it's still used in the environment.
With this, we ensure the Gardener operator is not blocked in upgrading to a newer version.

The constraints are used as a mechanism for reporting which shoot clusters still use the old configuration of API server proxy and shoot VPN client.

> This helps the Gardener operator to determine easily whether the corresponding feature gate be enabled.

Reviewer: ```suggestion
This helps the Gardener operator to determine easily whether the corresponding feature gate can be enabled.
```

The `hostname` field is sent in the CONNECT request to the upstream proxy (ingress gateway).
This value is discarded, but required in the config.
Additionally, the custom header `X-Gardener-Destination` is added to the CONNECT request to indicate the destination API server.
We don't reuse the `Reversed-VPN` header which fulfills the same purpose to avoid confusion.
However, the value of the `X-Gardener-Destination` header follows the same format (i.e., the Envoy cluster string format):

```text
outbound|443||kube-apiserver.<technical-shoot-id>.svc.cluster.local
```

In the corresponding upstream cluster section of the config, the destination port to connect to on istio-ingressgateway is changed from 8443 (`proxy`) to 8132 (`tls-tunnel`).
This configures the API server proxy to reuse the existing VPN infrastructure on the seed side.
Finally, the `transport_socket` is removed which disables adding the proxy protocol headers.

With this, a connection is established as follows:

> 1. An in-cluster client (e.g., pod) opens a TLS connection to `kubernetes.default.svc.cluster.local`. This service domain name points to an address which the API Server proxy listens on.

Reviewer: Nit:

```suggestion
1. An in-cluster client (e.g., pod) opens a TLS connection to `kubernetes.default.svc.cluster.local`. This service domain name points to an address which the API server proxy listens on.
```

I.e., we cannot assume that the port in question is no longer used, because the shoot reconciliation flow might not have reached the relevant tasks.
This might be because the shoot is marked to be ignored or an `Extension` with lifecycle `BeforeKubeAPIServer` cannot be reconciled successfully.
The feature gate allows the Gardener operator to delay the removal of the ingress gateway port if it's still used in the environment.
With this, we ensure the Gardener operator is not blocked in upgrading to a newer version.

The constraints are used as a mechanism for reporting which shoot clusters still use the old configuration of API server proxy and shoot VPN client.
This helps the Gardener operator to determine easily whether the corresponding feature gate can be enabled.
Additionally, gardenlet can prevent misconfiguration by iterating the relevant shoot constraints.

## Alternatives

We are not aware of any other alternative solution to address this issue, as requiring a transparent LoadBalancer for Gardener is no solution.

> ## Appendix

Reviewer 2: I'm tempted to drop the diagrams.
I feel like it's hard to visualize the architecture accurately while still making it helpful for understanding the topic.
I would hope that the precise descriptions in the document are good enough to understand the proposal.

WDYT?

Reviewer: Generally, visualisations can support describing the architecture. Therefore, I am usually all for having diagrams.
However, I can see why it is hard in this case to find the right visualisation. Having good diagrams is great, but getting there might be time-consuming.
Unless you have a good idea how to visualise it, feel free to drop them.

TLDR: I am fine either way.

Reviewer 2: I gave this a shot and reworked the diagrams to visualize the layered connections.
The diagrams are not perfectly accurate (layer numbers, destination IPs, header values, etc.), but they should be good enough to make the proposal easier to understand.

Looking forward to your feedback :)
